packages/cli/src/lib/bundler/config.ts

```diff
@@ -128,18 +129,81 @@ export async function createConfig(
     }),
   );
 
-  plugins.push(
-    new HtmlWebpackPlugin({
-      meta: {
-        'backstage-app-mode': options?.appMode ?? 'public',
-      },
-      template: paths.targetHtml,
-      templateParameters: {
-        publicPath,
-        config: frontendConfig,
-      },
-    }),
-  );
+  if (options.moduleFederation?.mode !== 'remote') {
+    plugins.push(
+      new HtmlWebpackPlugin({
+        meta: {
+          'backstage-app-mode': options?.appMode ?? 'public',
+        },
+        template: paths.targetHtml,
+        templateParameters: {
+          publicPath,
+          config: frontendConfig,
+        },
+      }),
+    );
+  }
+
+  if (options.moduleFederation) {
+    plugins.push(
+      new ModuleFederationPlugin({
+        ...(options.moduleFederation?.mode === 'remote' && {
+          filename: 'remoteEntry.js',
+          exposes: {
+            '.': paths.targetEntry,
+          },
+        }),
+        name: options.moduleFederation.name,
+        runtime: false,
+        shared: {
+          // React
+          react: {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          'react-dom': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          // React Router
+          'react-router': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          'react-router-dom': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          // MUI v5
+          '@material-ui/core/styles': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          '@material-ui/styles': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          // MUI v5
+          '@mui/material/styles/': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
+          },
+          '@emotion/react': {
+            singleton: true,
+            requiredVersion: '*',
+            eager: true,
```

Comment:
eager loading will always include packages in the entry chunk even if they were already loaded in the browser. This will make the init JS files very large.

Reply:
Ah of course! Though I based this on some earlier discussions but I think I had forgotten that we ended up not using it. What do you think about https://github.com/backstage/backstage/pull/25629/commits/858297c556b2363d03896ffc3b8bb46bfa90b0f3 though? Allowing these to be eager simplifies the app setup quite a bit, and including them as such in the host shouldn't matter much.

Reply:
Yeah for the shell app it does not matter as you have to load it anyway. Especially the singleton packages.